### PR TITLE
Build deb client pkgs

### PR DESCRIPTION
**Pull request by sbernhard**
Prepare jenkins jobs to build e.g. https://github.com/theforeman/foreman-packaging/pull/6958

### CODE REVIEW

**Comment by evgeni** [comment]
Are client bits OS independent, so it's sufficient to build them once? Given python is involved, I'd think not?

**Comment by sbernhard** [comment]
> Are client bits OS independent, so it's sufficient to build them once? Given python is involved, I'd think not?

well, they have dependencies on python library. If they are dependent on the OS, what do I need to consider in this PR?

**Comment by ekohl** [comment]
We need to decide on a layout. With https://github.com/theforeman/foreman-packaging/tree/deb/develop we have 2 layouts:
* OS dependent (`debian` and `dependencies`) which builds a package for each OS
* OS independent (`plugins`) which builds a package on one OS and assumes it works on all

Generally speaking, OS dependent is a safe assumption. In this case Python is very OS dependent since the version changes for every release.

In this PR it would be much safer to copy the `debian` / `dependencies` layout than `plugin`. I'd go for `dependencies` since it's the most straight forward. There are no nightly packages.

You could even go a step further and create a common function that's used in `setup_sources_dependency` and `setup_sources_client` to avoid duplicating everything, but I'm not sure how feasible this is.

**Comment by sbernhard** [comment]
Have a look again @ekohl , please :-)

File: theforeman.org/pipelines/lib/packaging.groovy [hunk 1]
```diff
@@ -74,6 +75,17 @@ def find_changed_debs(diff_range) {
...
     }
 
+    for(client in changed_client) {
+        (folder, project) = client.split('/')
+
+        changed_debs.add([
+            type: 'client',
+            name: project,
+            path: "${folder}/${project}",
+            operating_system: 'buster'
```
[the remark this answers is not in the document]
Comment (sbernhard) [reply theforeman.org/pipelines/lib/packaging.groovy]: Thank you very much @ekohl for your suggestions!

File: theforeman.org/pipelines/lib/packaging.groovy [hunk 2]
```diff
@@ -40,6 +40,7 @@ def find_changed_debs(diff_range) {
     def changed_dependencies = find_added_or_changed_files(diff_range, 'dependencies/*/*/changelog').split()
     def changed_plugins = find_added_or_changed_files(diff_range, 'plugins/*/changelog').split()
     def changed_core = find_added_or_changed_files(diff_range, 'debian/*/*/changelog').split()
+    def changed_client = find_added_or_changed_files(diff_range, 'client/*/changelog').split()
```
[the remark this answers is not in the document]
Comment (sbernhard) [reply theforeman.org/pipelines/lib/packaging.groovy]: Thanks.

**Review by ekohl — APPROVED** [review]
I think this looks good.

@evgeni could you also have a look?

Also many thanks to @ehelms for rewriting the pipeline, making this so much easier than it used to be.

**Review by evgeni — CHANGES_REQUESTED** [review]
this goes into the right direction, but is not yet sufficient!

with the current structure of the "publishing" workflow, this will land in the "core" repository, not in a dedicated "client" repo, which I think is the intention here?

https://github.com/theforeman/jenkins-jobs/blob/47c37530ed3d6d63d6d1a8bcdcddadcfee56b90a/theforeman.org/pipelines/lib/packaging.groovy#L141-L161

Right now, we have the following repo structure
```
deb http://deb.theforeman.org/ <OS> <version>
deb http://deb.theforeman.org/ plugins <version>
deb http://stagingdeb.theforeman.org/ <OS> <USER>-<version>
```

To translate:
* a core/dependencies PR gets published to `http://stagingdeb.theforeman.org/ <OS> theforeman-<version>` on merge
* a plugins PR gets published to `deb http://deb.theforeman.org/ plugins <version>` on merge
* when pipelines pass, stuff from `http://stagingdeb.theforeman.org/ <OS> theforeman-<version>` is copied over to `http://deb.theforeman.org/ <OS> <version>`

I think we should not stage client stuff (we have no tests for them anyways, and we don't do it right now for RPM either), but make it publish to `deb http://deb.theforeman.org/ <OS> <version>-client` directly. If you agree, I can propose a patch to :arrow_up: to do the same.

**Comment by evgeni** [comment]
I think the following should do it:

```patch
diff --git theforeman.org/pipelines/lib/packaging.groovy theforeman.org/pipelines/lib/packaging.groovy
index f5acc9a..ba49378 100644
--- theforeman.org/pipelines/lib/packaging.groovy
+++ theforeman.org/pipelines/lib/packaging.groovy
@@ -160,6 +160,16 @@ def build_deb_package_steps(packages_to_build, version, repoowner = 'theforeman'
             component = repoowner
             release_type = 'stage'
             echo "scratch build: uploading to stagingdeb/${suite}/${component}"
+        } else if (type == 'client' && !pull_request) {
+            suite = "${os}-client"
+            component = version
+            release_type = 'release'
+            echo "plugin build without PR: uploading directly to deb/${suite}/${component}"
+        } else if (type == 'client' && pull_request) {
+            suite = "${os}-client"
+            component = "${repoowner}-${version}"
+            release_type = 'stage'
+            echo "scratch build: uploading to stagingdeb/${suite}/${component}"
         } else {
             suite = os
             component = "${repoowner}-${version}"
```

**Comment by evgeni** [comment]
Or:

```
diff --git theforeman.org/pipelines/lib/packaging.groovy theforeman.org/pipelines/lib/packaging.groovy
index f5acc9a..ba49378 100644
--- theforeman.org/pipelines/lib/packaging.groovy
+++ theforeman.org/pipelines/lib/packaging.groovy
@@ -160,6 +160,16 @@ def build_deb_package_steps(packages_to_build, version, repoowner = 'theforeman'
             component = repoowner
             release_type = 'stage'
             echo "scratch build: uploading to stagingdeb/${suite}/${component}"
+        } else if (type == 'client' && !pull_request) {
+            suite = os
+            component = "${version}-client"
+            release_type = 'release'
+            echo "plugin build without PR: uploading directly to deb/${suite}/${component}"
+        } else if (type == 'client' && pull_request) {
+            suite = os
+            component = "${repoowner}-${version}-client"
+            release_type = 'stage'
+            echo "scratch build: uploading to stagingdeb/${suite}/${component}"
         } else {
             suite = os
             component = "${repoowner}-${version}"
```

**Comment by ekohl** [comment]
The former patch results in
```
deb http://deb.theforeman.org/ <OS> <version>
deb http://deb.theforeman.org/ <OS>-client <version>
```
While the latter results in:
```
deb http://deb.theforeman.org/ <OS> <version> <version>-client
```

I think I prefer the latter.

**Comment by evgeni** [comment]
> I think I prefer the latter.

So do I.

@sbernhard what do you think?

**Review by ekohl — APPROVED** [review]
This looks good to me, but I missed something earlier.

**Comment by sbernhard** [comment]
Is there anything left before this can be merged @evgeni / @ekohl ?

**Comment by ekohl** [comment]
I must admit I mostly forgot about this due to the many other tasks. I'd be in favor of getting this live next week.

**Review by evgeni — APPROVED** [review]

**Pull request merged**